core-api/src/main/java/com/optimizely/ab/config/audience/match/AttributeMatch.java

```diff
@@ -0,0 +1,28 @@
+/**
+ *
+ *    Copyright 2018, Optimizely and contributors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package com.optimizely.ab.config.audience.match;
+
+abstract class AttributeMatch<T> implements Match {
+    T convert(Object o) {
+        try {
+            T rv = (T)o;
+            return rv;
+        } catch(java.lang.ClassCastException e) {
+            return null;
```

Comment:
Swallowing the error and returning `null` feels pretty weird to me. This causes subclasses' `eval(...)` methods to proceed blindly with a `null` value (in some cases producing `NullPointerException`s which they must know to catch) when the more straightforward, robust behavior might be for them to immediately return `null`.

Basically, I think the pattern smells.

As one alternative: what if we tried something like
```java
abstract class TypedMatch<T> implements Match {
    Boolean eval(Object otherValue) {
        try {
            T otherValueT = (T) otherValue;
        } catch(java.lang.ClassCastException e) {
            return null;
        }
        return eval(typedOtherValue);
    }

    Boolean eval(T otherValue);
}
```
and had our subclasses the new, typed abstract method instead?

(Or give up on having common, generic-powered type-checking code? But maybe that's just my dynamic typing background coming through 🤷)

Reply:
This is a great idea and I really like where you are going with it.  However, in Java (and other strongly typed languages) the eval here ends up with the same signature.  Meaning a value is both of type object and type t.  So, it becomes ambiguous.  :(

Reply:
Ah, interesting, thanks.